README.md
# Mopidy.js

[![Latest npm version](https://img.shields.io/npm/v/mopidy.svg?style=flat)](https://www.npmjs.org/package/mopidy)
[![Number of npm downloads](https://img.shields.io/npm/dm/mopidy.svg?style=flat)](https://www.npmjs.org/package/mopidy)
[![Build Status](https://img.shields.io/travis/mopidy/mopidy.js.svg?style=flat)](https://travis-ci.org/mopidy/mopidy.js)

Mopidy.js is a JavaScript library for controlling a
[Mopidy music server](https://www.mopidy.com/).

The library makes
[Mopidy's core API](https://docs.mopidy.com/en/latest/api/core/)
available from browsers and Node.js programs,
using JSON-RPC over a WebSocket to communicate with Mopidy.

This library is the foundation of most Mopidy web clients.

## Table of contents <!-- omit in toc -->

- [Compatibility](#compatibility)
- [Installation](#installation)
  - [From npm](#from-npm)
  - [Prebuilt bundle](#prebuilt-bundle)
  - [Building from source](#building-from-source)
- [Usage](#usage)
  - [Importing](#importing)
  - [Creating an instance](#creating-an-instance)
  - [Settings](#settings)
    - [`autoConnect`](#autoconnect)
    - [`backoffDelayMin`](#backoffdelaymin)
    - [`backoffDelayMax`](#backoffdelaymax)
    - [`console`](#console)
    - [`webSocket`](#websocket)
    - [`webSocketUrl`](#websocketurl)
  - [Hooking up to events](#hooking-up-to-events)
    - [Client state](#client-state)
    - [Mopidy events](#mopidy-events)
    - [WebSocket events](#websocket-events)
  - [Calling core API methods](#calling-core-api-methods)
    - [API discovery](#api-discovery)
    - [Arguments](#arguments)
    - [Promises](#promises)
    - [async/await](#asyncawait)
  - [Cleaning up](#cleaning-up)
- [Demos](#demos)
  - [Web application](#web-application)
  - [Console application](#console-application)
- [Changelog](#changelog)

## Compatibility

Mopidy.js should work with Mopidy >= 0.19.0.

## Installation

Mopidy.js is a tiny library, building upon the standardized WebSocket and
Promise APIs, weighing in at around 12 kB minified.

### From npm

You can install Mopidy.js from npm:

```
npm install mopidy
```

### Prebuilt bundle

A minified versions of Mopidy.js, complete with sourcemaps, is available from
the project's
[GitHub release page](https://github.com/mopidy/mopidy.js/releases).

### Building from source

Install [Node.js](https://nodejs.org/).

Enter the source directory, and install all dependencies:

```
npm install
```

That's it.

You can now run the tests and linters:

```
npm test
```

To build updated JavaScript files for browser use in `dist/`, run:

```
npm run build
```

## Usage

### Importing

After installing, you can import Mopidy.js into your code using `require()`:

```js
const Mopidy = require("mopidy");
```

Or using ES6 imports:

```js
import Mopidy from "mopidy";
```

### Creating an instance

Once you have Mopidy.js loaded, you need to create an instance of the wrapper:

```js
const mopidy = new Mopidy();
```

When you instantiate `Mopidy()` without arguments, it will connect to
the WebSocket at `/mopidy/ws/` on the current host. Thus, if you don't host
your web client using Mopidy's built-in web server, or if you use Mopidy.js
from a Node.js app, you'll need to pass the URL to the WebSocket end point:

```js
const mopidy = new Mopidy({
  webSocketUrl: "ws://localhost:6680/mopidy/ws/",
});
```

It is also possible to create an instance first and connect to the WebSocket
later:

```js
const mopidy = new Mopidy({ autoConnect: false });
// ... do other stuff, like hooking up events ...
mopidy.connect();
```

### Settings

When creating an instance, you can specify the following settings:

#### `autoConnect`

Whether or not to connect to the WebSocket on instance creation.
Defaults to `true`.

#### `backoffDelayMin`

The minimum number of milliseconds to wait after a connection error before
we try to reconnect. For every failed attempt, the backoff delay is doubled
until it reaches `backoffDelayMax`. Defaults to `1000`.

#### `backoffDelayMax`

The maximum number of milliseconds to wait after a connection error before
we try to reconnect. Defaults to `64000`.

#### `console`

If set, this object will be used to log errors from Mopidy.js. This is
mostly useful for testing Mopidy.js. Defaults to `console`.

#### `webSocket`

An existing WebSocket object to be used instead of creating a new
WebSocket. Defaults to `undefined`.

#### `webSocketUrl`

URL used when creating new WebSocket objects.

In a browser environment, it defaults to
`ws://${document.location.host}/mopidy/ws`. If the current page is served over
HTTPS, it defaults to using `wss://` instead of `ws://`.

In a non-browser environment, where `document.location` isn't
available, it defaults to `ws://localhost/mopidy/ws`.

### Hooking up to events

Once you have a Mopidy.js object, you can hook up to the events it emits. To
explore your possibilities, it can be useful to subscribe to all events and log
them:

```js
mopidy.on("state", console.log);
mopidy.on("event", console.log);
```

Mopidy.js uses [Node.js' events API](https://nodejs.org/api/events.html), so
you should refer to its documentation to see how you can hook up your listeners
to the different events.

Several types of events are emitted:

#### Client state

You can get notified about when the Mopidy.js client is connected to the
server and ready for method calls, when it's offline, and when it's trying to
reconnect to the server by looking at the events `state:online`,
`state:offline`, `reconnectionPending`, and `reconnecting`.

The client state events are also emitted under the aggregate event named
`state`.

#### Mopidy events

You can get events sent from the Mopidy server by looking at the events with
the name prefix `event:`, like `event:trackPlaybackStarted`.

The events from Mopidy are also emitted under the aggregate event named
`event`.

#### WebSocket events

You can introspect what happens internally on the WebSocket by looking at the
events named `websocket:open`, `websocket:error`, `websocket:close`, `websocket:incomingMessage`, and `websocket:outgoingMessage`.

Of course, you can also do this using the web developer tools in any modern browser.

### Calling core API methods

Once your Mopidy.js object has connected to the Mopidy server and emits the
`state:online` event, it is ready to accept core API method calls:

```js
mopidy.on("state:online", () => {
  mopidy.playback.next();
});
```

Any calls you make before the `state:online` event is emitted will fail with a
`Mopidy.ConnectionError` error.

All methods in [Mopidy's core API](https://docs.mopidy.com/en/latest/api/core/)
is available via Mopidy.js. For example, the
`mopidy.core.PlaybackController.get_state()` Python method is available
in JSON-RPC as `core.playback.get_state` and in Mopidy.js as
`mopidy.playback.getState()`.

Both the WebSocket API and the JavaScript API are based on introspection of the
core Python API. Thus, they will always be up to date and immediately reflect
any changes we do to the core API.

#### API discovery

The best way to explore the JavaScript API, is probably by opening your
browser's console, and using its tab completion to navigate the API. You'll
find the Mopidy core API exposed under:

- `mopidy.playback`
- `mopidy.tracklist`
- `mopidy.library`
- `mopidy.playlists`
- `mopidy.history`
- `mopidy.mixer`

All methods in the JavaScript API have an associated data structure describing
the Python params it expects, and most methods also have the Python API
documentation available. This is available right there in the browser console,
by looking at the method's `description` and `params` attributes:

```js
console.log(mopidy.playback.next.params);
console.log(mopidy.playback.next.description);
```

#### Arguments

JSON-RPC 2.0 limits method parameters to be sent _either_ by-position or
by-name. Combinations of both, like we're used to from Python, isn't supported
by JSON-RPC 2.0.

To call a method with by-position arguments, wrap the arguments in a list:

```js
mopidy.library.search(["abba", null, true]);
```

As you can see, this quickly becomes confusing as we don't know what `null`
and `true` is without referring to documentation or Python source code.

Thus, it is usually better to call methods with by-name arguments, by passing
an object to the method:

```js
mopidy.library.search({ query: "abba", exact: true });
```

#### Promises

Obviously, you'll want to get a return value from many of your method calls.
Since everything is happening across the WebSocket and maybe even across the
network, you'll get the results asynchronously. Instead of having to pass
callbacks and errbacks to every method you call, the methods return
ES6 `Promise` objects, which you can use to pipe the future result as input
to another method, or to hook up callback and errback functions.

```js
const track = mopidy.playback.getCurrentTrack();
// => `track` isn't an object with track data, but a Promise
```

Typical usage of a `Promise` can look like this:

```js
function showPlaying() {
  const printCurrentTrack = (track) => {
    if (track) {
      console.log(
        "Currently playing:",
        track.name,
        "by",
        track.artists[0].name,
        "from",
        track.album.name
      );
    } else {
      console.log("No current track");
    }
  };

  const failureHandler = () => {
    console.warn("Something went wrong");
  };

  mopidy.playback.getCurrentTrack().then(printCurrentTrack, failureHandler);
}
```

The first function passed to `.then()`, `printCurrentTrack`, is the callback
that will be called if the method call succeeds. If anything goes wrong,
the second function passed to `.then()`, `failureHandler` will be called.

#### async/await

With the foundation of `Promise`, you can simplify the code flow using the
`async` and `await` keywords introduced to JavaScript in ES2017.

The following is the same example written using `async` and `await`:

```js
async function showPlaying() {
  try {
    const track = await mopidy.playback.getCurrentTrack();
    if (track) {
      console.log(
        "Currently playing:",
        track.name,
        "by",
        track.artists[0].name,
        "from",
        track.album.name
      );
    } else {
      console.log("No current track");
    }
  } catch (e) {
    console.warn("Something went wrong", e);
  }
}
```

### Cleaning up

If you for some reason want to clean up after Mopidy.js before the web page is
closed or navigated away from, you can close the WebSocket, unsubscribe all
event listeners, and delete the object like this:

```js
// Close the WebSocket without reconnecting. Letting the object be garbage
// collected will have the same effect, so this isn't strictly necessary.
mopidy.close();

// Unsubscribe all event listeners. If you don't do this, you may have
// lingering references to the object causing the garbage collector to not
// clean up after it.
mopidy.off();

// Delete your reference to the object, so it can be garbage collected.
mopidy = null;
```

## Demos

### Web application

![Web demo](examples/web.png)

In the `examples/` directory of the Git repo, you can find a small demo web
application using Mopidy.js. The left half of the screen shows what's
currently playing and provides some basic playback controls. The right half of
the screen shows the JSON-RPC messages and events that are sent back and forth
to the server, hopefully giving some insight into what is available to
Mopidy.js developers.

To run the demo application yourself:

1. Make sure the `http/allowed_origins` config value in your `mopidy.conf`
   includes `localhost:1234`.

2. Run Mopidy on your local machine, so that Mopidy's web interface becomes
   available at http://localhost:6680/.

3. Clone Mopidy.js from GitHub.

4. Run `npm install` to install dependencies.

5. Run `npm start` to run the demo application at http://localhost:1234/.

This setup uses hot module reloading, so any changes you do to the demo
application files, `examples/web.{html,js}`, will instantly be visible in
your browser. Thus, this can serve as a nice playing ground to get to know
the capabilities of Mopidy and Mopidy.js.

### Console application

In the `examples/` directory of the Git repo, you can find `mpc.js`, a partial
clone of the `mpc` utility built using Mopidy.js on Node.

Output from the original command:

```
$ mpc
Jon Hopkins - C O S M
[playing] #6/12   2:33/7:08 (35%)
volume:100%   repeat: off   random: on    single: off   consume: off
```

Output from the `mpc.js` example:

```
$ ./examples/mpc.js
Jon Hopkins - C O S M
[playing] #6/-   2:34/7:08 (36%)
volume:100%   repeat: off   random: on    single: off   consume: off
```

To run this example yourself, follow step 1-4 for the demo web application, and
then run `./examples/mpc.js`.

## Changelog

See [CHANGELOG.md](CHANGELOG.md).


CHANGELOG.md
# Changelog

## [1.2.1] - 2020-11-21

- Include TypeScript definitions in npm package.

## [1.2.0] - 2020-11-21

- Add TypeScript definitions. (Fixes: #27, #33)

## [1.1.1] - 2020-09-10

- Replace `setImmediate(fn)` usage with `setTimeout(fn, 0)` as
  `setImmediate()` is non-standard and not supported in all browsers.
  (Fixes: #30)

## [1.1.0] - 2020-04-02

- More sensible event ordering: `state:offline` is now emitted before
  `reconnectionPending`. (Contributed by Alan Norbauer. PR #24)

- Upgrade all development and production dependencies.

## [1.0.1] - 2019-04-04

- Fix `TypeError` on failure to get the API spec. (Contributed by Gamadril.
  PR: #19)

## [1.0.0] - 2018-12-01

- The prebuilt files at the
  [GitHub release page](https://github.com/mopidy/mopidy.js/releases)
  now includes a source map instead of an uncompressed version.

- **Backwards incompatible:** The `Mopidy` class can no longer be instantiated
  without the `new` keyword.

  To upgrade existing code:

  ```js
  // Change from this:
  const mopidy = Mopidy(...);
  // To this:
  const mopidy = new Mopidy(...);
  ```

- **Backwards incompatible:** The API methods no longer support two different
  calling conventions. (Fixes: #10)

  To upgrade code using `by-position-or-by-name`, simply remove
  `callingConvention` from the settings object:

  ```js
  // Change from this:
  const mopidy = new Mopidy({ callingConvention: "by-position-or-by-name" });
  // To this:
  const mopidy = new Mopidy();
  ```

  To upgrade code using the long deprecated, but still default, calling
  convention `by-position-only` multiple steps are required.

  1.  The first step is to remove `callingConvention` from the settings object:

      ```js
      // Change from this:
      const mopidy = new Mopidy({ callingConvention: "by-position-only" });
      // To this:
      const mopidy = new Mopidy();
      ```

  2.  The second step is to update all API method calls to explicitly use
      positional arguments:

      ```js
      // Change from this:
      mopidy.tracklist.setRepeat(true);
      // To this:
      mopidy.tracklist.setRepeat([true]);
      ```

      At this point, the application should work again.

  3.  The final, and optional step, is to change API method calls to using objects
      instead of arrays where that makes the code clearer:

      ```js
      // Optionally change from this:
      mopidy.library.search(["abba", null, true]);
      // To this:
      mopidy.library.search({ query: "abba", exact: true });
      ```

- **Backwards incompatible:** The `Mopidy` class no longer reexports When.js
  as `Mopidy.when()`. To upgrade existing code, either migrate to standard
  ES6 `Promise` or add When.js as a dependency to your project.

- **Backwards incompatible:** The event listening API has changed a bit as
  Mopidy.js has switched from the unmaintained BANE implementation to Node.js'
  [`EventEmitter` API](https://nodejs.org/api/events.html).

  The following methods have been removed:

  - `mopidy.on(listener)` --
    Subscribe to more specific events using `mopidy.on(event, listener)`.
  - `mopidy.off(listener)` --
    Unsubscribe from specific events using `mopidy.off(event, listener)`.
  - `mopidy.bind(...)`
  - `mopidy.errback(listener)`

  The following methods works as before:

  - `mopidy.on(event, listener)`
  - `mopidy.once(event, listener)`
  - `mopidy.off(event, listener)`
  - `mopidy.off(event)`
  - `mopidy.off()`
  - `mopidy.emit(event, ...)`

- For exploring what events Mopidy.js emit, two new aggregate event types
  have been added:

  - `state`: This event emits online/offline/reconnect events,
    like `state:online`.
  - `event`: This event emits server side events from Mopidy,
    like `event:trackPlaybackStarted`.

  Listeners to the aggregate events get arguments on the form
  `(eventName, data)`, where `eventName` is the name of the more specific
  event. It is recommended that applications do not use these aggregates, but
  instead subscribe to more specific events.

- Fixed `TypeError` when calling `mopidy.close()` when no WebSocket connection
  exists, for example when using `new Mopidy({ autoConnect: false })`.
  (Fixes: #15)

- Modernized dependencies:

  - The `Promise` object standardized in ES6 has replaced When.js. (Fixes: #11)
  - The `EventEmitter` object from Node.js, with polyfills for the browser,
    has replaced BANE. (Fixes: #12)
  - `isomorphic-ws` and `ws` has replaced our own wrapper around the browser's
    `WebSocket` API and `faye-websocket` on Node.

- Modernized development stack:

  - Testing: Jest has replaced Buster.JS and Sinon. (Fixes: #13)
  - Linting: ESLint has replaced JSHint.
  - Building: Parcel has replaced Browserify and Uglify.
  - Tasks: npm scripts have replaced Grunt.
  - Formatting: Prettier has replaced manual formatting.

## [0.5.0] - 2015-01-31

- Reexport When.js library as `Mopidy.when`, to make it easily available to
  users of Mopidy.js. (Fixes: #1)

- Default to `wss://` as the WebSocket protocol if the page is hosted on
  `https://`. This has no effect if the `webSocketUrl` setting is specified.
  (PR: #2)

- Upgrade dependencies.

## [0.4.1] - 2014-09-11

- Update links to point to new independent Mopidy.js GitHub project.

## [0.4.0] - 2014-06-24

- Add support for method calls with by-name arguments. The old calling
  convention, "by-position-only", is still the default, but this will change in
  the future. A warning is printed to the console if you don't explicitly
  select a calling convention. See the docs for details.

## [0.3.0] - 2014-06-16

- Upgrade to when.js 3, which brings great performance improvements and better
  debugging facilities. If you maintain a Mopidy client, you should review the
  [differences between when.js 2 and 3](https://github.com/cujojs/when/blob/master/docs/api.md#upgrading-to-30-from-2x)
  and the
  [when.js debugging guide](https://github.com/cujojs/when/blob/master/docs/api.md#debugging-promises).


src/mopidy.js
const EventEmitter = require("events");
const WebSocket = require("isomorphic-ws");

function snakeToCamel(name) {
  return name.replace(/(_[a-z])/g, (match) =>
    match.toUpperCase().replace("_", "")
  );
}

class Mopidy extends EventEmitter {
  constructor(settings) {
    super();
    this._console = this._getConsole(settings || {});
    this._settings = this._configure(settings || {});
    this._backoffDelay = this._settings.backoffDelayMin;
    this._pendingRequests = {};
    this._webSocket = null;
    this._delegateEvents();
    if (this._settings.autoConnect) {
      this.connect();
    }
  }

  _getConsole(settings) {
    if (typeof settings.console !== "undefined") {
      return settings.console;
    }
    const con = (typeof console !== "undefined" && console) || {};
    con.log = con.log || (() => {});
    con.warn = con.warn || (() => {});
    con.error = con.error || (() => {});
    return con;
  }

  _configure(settings) {
    const newSettings = { ...settings };
    const protocol =
      typeof document !== "undefined" && document.location.protocol === "https:"
        ? "wss://"
        : "ws://";
    const currentHost =
      (typeof document !== "undefined" && document.location.host) ||
      "localhost";
    newSettings.webSocketUrl =
      settings.webSocketUrl || `${protocol}${currentHost}/mopidy/ws`;
    if (settings.autoConnect !== false) {
      newSettings.autoConnect = true;
    }
    newSettings.backoffDelayMin = settings.backoffDelayMin || 1000;
    newSettings.backoffDelayMax = settings.backoffDelayMax || 64000;
    return newSettings;
  }

  _delegateEvents() {
    // Remove existing event handlers
    this.removeAllListeners("websocket:close");
    this.removeAllListeners("websocket:error");
    this.removeAllListeners("websocket:incomingMessage");
    this.removeAllListeners("websocket:open");
    this.removeAllListeners("state:offline");
    // Register basic set of event handlers
    this.on("websocket:close", this._cleanup);
    this.on("websocket:error", this._handleWebSocketError);
    this.on("websocket:incomingMessage", this._handleMessage);
    this.on("websocket:open", this._resetBackoffDelay);
    this.on("websocket:open", this._getApiSpec);
    this.on("state:offline", this._reconnect);
  }

  off(...args) {
    if (args.length === 0) {
      this.removeAllListeners();
    } else if (args.length === 1) {
      const arg = args[0];
      if (typeof arg === "string") {
        this.removeAllListeners(arg);
      } else {
        throw Error(
          "Expected no arguments, a string, or a string and a listener."
        );
      }
    } else {
      this.removeListener(...args);
    }
  }

  connect() {
    if (this._webSocket) {
      if (this._webSocket.readyState === Mopidy.WebSocket.OPEN) {
        return;
      }
      this._webSocket.close();
    }

    this._webSocket =
      this._settings.webSocket ||
      new Mopidy.WebSocket(this._settings.webSocketUrl);

    this._webSocket.onclose = (close) => {
      this.emit("websocket:close", close);
    };
    this._webSocket.onerror = (error) => {
      this.emit("websocket:error", error);
    };
    this._webSocket.onopen = () => {
      this.emit("websocket:open");
    };
    this._webSocket.onmessage = (message) => {
      this.emit("websocket:incomingMessage", message);
    };
  }

  _cleanup(closeEvent) {
    Object.keys(this._pendingRequests).forEach((requestId) => {
      const { reject } = this._pendingRequests[requestId];
      delete this._pendingRequests[requestId];
      const error = new Mopidy.ConnectionError("WebSocket closed");
      error.closeEvent = closeEvent;
      reject(error);
    });
    this.emit("state", "state:offline");
    this.emit("state:offline");
  }

  _reconnect() {
    // We asynchronously process the reconnect because we don't want to start
    // emitting "reconnectionPending" events before we've finished handling the
    // "state:offline" event, which would lead to emitting the events to
    // listeners in the wrong order.
    setTimeout(() => {
      this.emit("state", "reconnectionPending", {
        timeToAttempt: this._backoffDelay,
      });
      this.emit("reconnectionPending", {
        timeToAttempt: this._backoffDelay,
      });
      setTimeout(() => {
        this.emit("state", "reconnecting");
        this.emit("reconnecting");
        this.connect();
      }, this._backoffDelay);
      this._backoffDelay *= 2;
      if (this._backoffDelay > this._settings.backoffDelayMax) {
        this._backoffDelay = this._settings.backoffDelayMax;
      }
    }, 0);
  }

  _resetBackoffDelay() {
    this._backoffDelay = this._settings.backoffDelayMin;
  }

  close() {
    this.off("state:offline", this._reconnect);
    if (this._webSocket) {
      this._webSocket.close();
    }
  }

  _handleWebSocketError(error) {
    this._console.warn("WebSocket error:", error.stack || error);
  }

  _send(message) {
    switch (this._webSocket.readyState) {
      case Mopidy.WebSocket.CONNECTING:
        return Promise.reject(
          new Mopidy.ConnectionError("WebSocket is still connecting")
        );
      case Mopidy.WebSocket.CLOSING:
        return Promise.reject(
          new Mopidy.ConnectionError("WebSocket is closing")
        );
      case Mopidy.WebSocket.CLOSED:
        return Promise.reject(
          new Mopidy.ConnectionError("WebSocket is closed")
        );
      default:
        return new Promise((resolve, reject) => {
          const jsonRpcMessage = {
            ...message,
            jsonrpc: "2.0",
            id: this._nextRequestId(),
          };
          this._pendingRequests[jsonRpcMessage.id] = { resolve, reject };
          this._webSocket.send(JSON.stringify(jsonRpcMessage));
          this.emit("websocket:outgoingMessage", jsonRpcMessage);
        });
    }
  }

  _handleMessage(message) {
    try {
      const data = JSON.parse(message.data);
      if (Object.hasOwnProperty.call(data, "id")) {
        this._handleResponse(data);
      } else if (Object.hasOwnProperty.call(data, "event")) {
        this._handleEvent(data);
      } else {
        this._console.warn(
          `Unknown message type received. Message was: ${message.data}`
        );
      }
    } catch (error) {
      if (error instanceof SyntaxError) {
        this._console.warn(
          `WebSocket message parsing failed. Message was: ${message.data}`
        );
      } else {
        throw error;
      }
    }
  }

  _handleResponse(responseMessage) {
    if (
      !Object.hasOwnProperty.call(this._pendingRequests, responseMessage.id)
    ) {
      this._console.warn(
        "Unexpected response received. Message was:",
        responseMessage
      );
      return;
    }
    const { resolve, reject } = this._pendingRequests[responseMessage.id];
    delete this._pendingRequests[responseMessage.id];
    if (Object.hasOwnProperty.call(responseMessage, "result")) {
      resolve(responseMessage.result);
    } else if (Object.hasOwnProperty.call(responseMessage, "error")) {
      const error = new Mopidy.ServerError(responseMessage.error.message);
      error.code = responseMessage.error.code;
      error.data = responseMessage.error.data;
      reject(error);
      this._console.warn("Server returned error:", responseMessage.error);
    } else {
      const error = new Error("Response without 'result' or 'error' received");
      error.data = { response: responseMessage };
      reject(error);
      this._console.warn(
        "Response without 'result' or 'error' received. Message was:",
        responseMessage
      );
    }
  }

  _handleEvent(eventMessage) {
    const data = { ...eventMessage };
    delete data.event;
    const eventName = `event:${snakeToCamel(eventMessage.event)}`;
    this.emit("event", eventName, data);
    this.emit(eventName, data);
  }

  _getApiSpec() {
    return this._send({ method: "core.describe" })
      .then(this._createApi.bind(this))
      .catch(this._handleWebSocketError.bind(this));
  }

  _createApi(methods) {
    const caller = (method) => (...args) => {
      const message = { method };
      if (args.length === 0) {
        return this._send(message);
      }
      if (args.length > 1) {
        return Promise.reject(
          new Error(
            "Expected zero arguments, a single array, or a single object."
          )
        );
      }
      if (!Array.isArray(args[0]) && args[0] !== Object(args[0])) {
        return Promise.reject(new TypeError("Expected an array or an object."));
      }
      [message.params] = args;
      return this._send(message);
    };

    const getPath = (fullName) => {
      let path = fullName.split(".");
      if (path.length >= 1 && path[0] === "core") {
        path = path.slice(1);
      }
      return path;
    };

    const createObjects = (objPath) => {
      let parentObj = this;
      objPath.forEach((objName) => {
        const camelObjName = snakeToCamel(objName);
        parentObj[camelObjName] = parentObj[camelObjName] || {};
        parentObj = parentObj[camelObjName];
      });
      return parentObj;
    };

    const createMethod = (fullMethodName) => {
      const methodPath = getPath(fullMethodName);
      const methodName = snakeToCamel(methodPath.slice(-1)[0]);
      const object = createObjects(methodPath.slice(0, -1));
      object[methodName] = caller(fullMethodName);
      object[methodName].description = methods[fullMethodName].description;
      object[methodName].params = methods[fullMethodName].params;
    };

    Object.keys(methods).forEach(createMethod);

    this.emit("state", "state:online");
    this.emit("state:online");
  }
}

class ConnectionError extends Error {
  constructor(message) {
    super(message);
    this.name = "ConnectionError";
  }
}
Mopidy.ConnectionError = ConnectionError;

class ServerError extends Error {
  constructor(message) {
    super(message);
    this.name = "ServerError";
  }
}
Mopidy.ServerError = ServerError;

Mopidy.WebSocket = WebSocket;

Mopidy.prototype._nextRequestId = (() => {
  let lastUsed = -1;
  return () => {
    lastUsed += 1;
    return lastUsed;
  };
})();

module.exports = Mopidy;
